From 84480a8b640b021f19edfc4671bce55b373c47c2 Mon Sep 17 00:00:00 2001
From: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
Date: Sun, 5 Nov 2023 22:45:09 +0530
Subject: [PATCH] ath12k: napi node addition not required for dummy irq

napi nodes are added for dummy irq for PPE when we do rmmod only
upto ext_irq_grp_num_max node got removed from napi but this dummy
noes are not consider this leads to dangling pointer and crash
at next insmod.

As the napi node functionlity not used for dummy irq nodes
hence the dummy irq nodes are not added to napi this will avoid
dangling pointer

Crash Signature :
Unable to handle kernel paging request at virtual address 0035203220310919
During the crash,
PC points to "napi_by_id+0x28/0x3c"
LR points to "netif_napi_add_weight+0x180/0x204".
The Backtrace obtained is as follows:
napi_by_id+0x28/0x3c
ath12k_pcic_get_msi_data+0x1e8/0x418 [ath12k]
ath12k_pcic_config_hybrid_irq+0x24c/0x3b4 [ath12k]
ath12k_rx_stats_buf_pktlog_process+0x1964/0x1a9c [ath12k]
platform_probe+0x64/0xb4
really_probe+0x1bc/0x380
__driver_probe_device+0x130/0x148
driver_probe_device+0x3c/0xc0
__driver_attach+0x15c/0x16c
bus_for_each_dev+0x6c/0xb4
driver_attach+0x20/0x28

Signed-off-by: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/pcic.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index e266e8d..8809009 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -378,9 +378,6 @@ static int ath12k_pcic_ext_config_gic_msi_irq(struct ath12k_base *ab,
 	if (ab->hw_params->ring_mask->tx[i])
 		budget = tx_comp_budget;
 
-	netif_napi_add_weight(&irq_grp->napi_ndev, &irq_grp->napi,
-		       ath12k_pcic_ext_grp_napi_poll, budget);
-
 	if (ab->hw_params->ring_mask->tx[i] ||
 	    ab->hw_params->ring_mask->rx[i] ||
 	    ab->hw_params->ring_mask->rx_err[i] ||
@@ -415,6 +412,9 @@ static int ath12k_pcic_ext_config_gic_msi_irq(struct ath12k_base *ab,
 			}
 		} else {
 #endif
+			netif_napi_add_weight(&irq_grp->napi_ndev, &irq_grp->napi,
+		       ath12k_pcic_ext_grp_napi_poll, budget);
+
 			scnprintf(dp_irq_name[userpd_id][i], DP_IRQ_NAME_LEN,
 				  "pci%u_wlan_dp_%u", userpd_id, i);
 			irq_set_status_flags(msi_desc->irq, IRQ_DISABLE_UNLAZY);
-- 
2.17.1

